### PR TITLE
First pass at optional root element

### PR DESCRIPTION
**Pull request by designermonkey**
1. Adds ability to pass `el: false` via View.
2. Handles errors if both `template: false` and `el: false`.
3. Changes `attachElContent` in `ItemView` and `CompositeView` to allow replacing default `$el`.

This is my very naive first attempt at this. I have more than likely not considered some aspects, so please feel free to tell me where I've not looked.

One thing that jumped out at me is that `CompositeView` doesn't handle errors when `template` is false. Is that expected behaviour?

I am new to Marionette, so please help me get this PR up to scratch.

### CODE REVIEW

**Comment by designermonkey** [comment]
#2357

**Comment by samccone** [comment]
I would start with getting the tests to pass and adding some specs. We can review the implementation details after :)

**Comment by thejameskyle** [comment]
This is not a pattern we absolutely should not be changing

File: src/view.js [hunk 1]
```diff
@@ -17,6 +18,10 @@ Marionette.View = Backbone.View.extend({
     // at some point however this may be removed
     this.options = _.extend({}, _.result(this, 'options'), options);
 
+    if (this.options && ('el' in this.options ? this.options.el === false : this.el === false)) {
```
Comment (megawac) [inline src/view.js]: You can do `this.getOption('el') === false` here

**Comment by jmeas** [comment]
@tbranyen told me that he's happy with the way LayoutManager handles this. If we get his :+1: on an implementation then I am +1.

**Comment by samccone** [comment]
@thejameskyle I kinda want to see where this or goes. I view it as an exploration on top of the issue.  If you have comments regarding the feature I would voice your thoughts over on the origin issue

File: src/view.js (hunk 1 above)
Comment (designermonkey) [inline src/view.js]: Thanks for the hint, I'll update.

**Comment by designermonkey** [comment]
@samccone this is where I start to fail myself. I have no idea how testing in JS works I'm afraid, and honestly haven't got a lot of time to learn right now (way too much client work on). Would you be able to help?

@thejameskyle I can understand the reservation, but this would be an optional thing that wouldn't change any default behaviour.

**Comment by tbranyen** [comment]
@thejameskyle care to elaborate why? Especially given the absoluteness of your comment.  I don't know exactly how Marionette Views are architected, but the only issue we've faced w/ LayoutManager and optional root elements are when developers try and use two root elements, which would obviously not work and breaks the conventions in place w/ Backbone.  Element:View are 1:1.

It can totally work (and does).  We are very well tested for this feature so I have hope it'll make it into Marionette as well.

The only fear we had was that Backbone would change the semantics of `el` in the future, but that was a risk we were willing to take, believing naively in semver.  Recently though, I've grown more and more concerned, since it seems like Backbone could break this in the future at a patch level.

**Comment by designermonkey** [comment]
I have been naive here, as I've completely overlooked how `CollectionView` works. Sorry. I don't know enough to make this work without guidance.

**Comment by thejameskyle** [comment]
The element to view pattern is one of the very few contracts you have with any Backbone View. From the moment it's created until destroyed a View has a single element that everything can be based upon, from event delegation to a lot of Marionette's features like the `ui` hash. This makes it fairly simple to author mixins, plugins, behaviors, etc. There needs to be a _really_ good reason to break this contract. The `<dl>` case is unfortunate but I'm not sure that's a good enough reason to make this change.

**Comment by thejameskyle** [comment]
Also, this is pretty common in other frameworks' views/components. Both Ember and React need a root element. I believe angular directives need some kind of root node as well (I believe this can be any type of node?).

**Comment by jmeas** [comment]
Random thought: I wonder if a document fragment could do some magic here. Imagine if a user specified that they didn't want an element, and their template contains many root nodes. What if we rendered it into a document fragment, and did all of the event delegation / ui binding / etc. there. By keeping a reference to the fragment around, we could manage everything like normal. When it comes time to append it to the DOM, the fragment would just vanish.

I admit I'm not sure how much of the HTMLElement API fragments have tho' / how well they work with jquery / and so on. But I'd like to investigate the possibility!

**Comment by tbranyen** [comment]
@thejameskyle both Ember and React enforce the rules trying to be added by this PR.  With React you don't define the root element on the class definition, you define it in the template.  This allows you to do much more modifications to the root element via the template syntax instead of programmatically in the instance methods.

In the case of LayoutManager we warn the user when they use two top level elements, and React has done the same.  It's a workable pattern.

**Comment by thejameskyle** [comment]
@tbranyen I meant in the sense that they enforce that rule, not that it's defined in the same way. I don't think a view should have multiple root elements and I `this.($)el` should always refer to that single root element.

**Comment by jasonLaster** [comment]
Closing this PR as it stale and I don't see the idea moving along. 

Thankyou for the work @designermonkey and really good discussionguys @thejameskyle, @tbranyen, @thejameskyle.

**Pull request closed without merging**